app/frontend/src/components/EditLocationMap.tsx

```diff
@@ -55,136 +74,251 @@ export default function EditLocationMap({
-    // TODO: better default?
-    location
-      ? new LngLat(location.lng, location.lat)
-      : new LngLat(151.2099, -33.865143)
+    location ? new LngLat(rLocation.lng, rLocation.lat) : new LngLat(35, 10)
   );
+  const radius = useRef<number>(rLocation.radius);
 
-  const onMouseDown = (e: MapMouseEvent | MapTouchEvent) => {
+  const onCircleMouseDown = (e: MapMouseEvent | MapTouchEvent) => {
     // Prevent the default map drag behavior.
     e.preventDefault();
 
     map.current!.getCanvas().style.cursor = "grab";
 
     if (e.type === "touchstart") {
-      const handleTouchMove = (e: MapTouchEvent) => onMove(e);
+      const handleTouchMove = (e: MapTouchEvent) => onCircleMove(e);
       map.current!.on("touchmove", handleTouchMove);
-      map.current!.once("touchend", (e) => onUp(e, handleTouchMove));
+      map.current!.once("touchend", (e) => onCircleUp(e, handleTouchMove));
     } else {
-      const handleMove = (e: MapMouseEvent) => onMove(e);
+      const handleMove = (e: MapMouseEvent) => onCircleMove(e);
       map.current!.on("mousemove", handleMove);
-      map.current!.once("mouseup", (e) => onUp(e, handleMove));
+      map.current!.once("mouseup", (e) => onCircleUp(e, handleMove));
     }
   };
 
-  const onMove = (e: MapMouseEvent | MapTouchEvent) => {
-    centerCoords.current = e.lngLat.wrap();
-
-    (map.current!.getSource("location") as GeoJSONSource).setData(
-      pointGeoJson(centerCoords.current!)
-    );
+  const onCircleMove = (e: MapMouseEvent | MapTouchEvent) => {
+    centerCoords.current = e.lngLat;
+    redrawMap();
   };
 
-  const onUp = (
+  const onCircleUp = (
     e: MapMouseEvent | MapTouchEvent,
-    moveEvent: (x: any) => void
+    moveHandler: (x: any) => void
   ) => {
-    map.current!.off("mousemove", moveEvent);
-    map.current!.off("touchmove", moveEvent);
+    map.current!.off("mousemove", moveHandler);
+    map.current!.off("touchmove", moveHandler);
     map.current!.getCanvas().style.cursor = "move";
 
-    onMove(e);
-    setLocation({
-      lat: centerCoords.current!.lat,
-      lng: centerCoords.current!.lng,
+    centerCoords.current = e.lngLat;
+    redrawMap();
+    syncCoords();
+  };
+
+  const redrawMap = () => {
+    if (!exact) {
+      (map.current!.getSource("circle") as GeoJSONSource).setData(
+        circleGeoJson(centerCoords.current!, radius.current)
+      );
+    } else {
+      (map.current!.getSource("circle") as GeoJSONSource).setData(
+        pointGeoJson(centerCoords.current!)
+      );
+    }
+  };
+
+  // syncs imperative coordinates into the reactive location object
+  const syncCoords = () => {
+    const wrapped = centerCoords.current!.wrap();
+    commit((rLocation) => {
+      return {
+        ...rLocation,
+        lat: wrapped.lat,
+        lng: wrapped.lng,
+      };
     });
   };
 
+  const commit = async (
+    update: (s: ApproximateLocation) => ApproximateLocation
+  ) => {
+    await setRLocation(update);
```

Comment:
remove async and await, has no effect here (setRLocaion doesn't return a promise)

Reply:
thought set state was async in react?

Reply:
It is but you have no control